src/write_statsd.c

```diff
@@ -0,0 +1,476 @@
+      free(configuration.event_line_format);
+      return -2;
+    }
+    ssnprintf(configuration.event_line_format, buffer_len,
+              "%s.%%s.%%s:%%s|%%s", configuration.prefix);
+    ssnprintf(configuration.short_event_line_format, buffer_len - 3,
+              "%s.%%s:%%s|%%s", configuration.prefix);
+
+  } else {
+    /* <prefix>.%s.%s.<potfix>:%s|%s && <prefix>.%s.<potfix>:%s|%s */
+    len = 5 + strlen(configuration.prefix) + strlen(configuration.postfix);
+    buffer_len = len + 8 + 1;
+    configuration.event_line_format = allocate(buffer_len);
+    if (configuration.event_line_format == NULL) {
+      return -2;
+    }
+    configuration.short_event_line_format = allocate(buffer_len - 3);
+    if (configuration.short_event_line_format == NULL) {
+      free(configuration.event_line_format);
+      return -2;
+    }
+    ssnprintf(configuration.event_line_format, buffer_len,
+              "%s.%%s.%%s.%s:%%s|%%s", configuration.prefix,
+              configuration.postfix);
+    ssnprintf(configuration.short_event_line_format, buffer_len - 3,
+              "%s.%%s.%s:%%s|%%s", configuration.prefix,
+              configuration.postfix);
+
+  }
+  configuration.event_line_base_len = len;
+  configuration.short_event_line_base_len = len - 1;
+
+  /* Postfix and prefix are now in event_line_format so they can be freed. */
+  FREE_NOT_NULL(configuration.postfix)
+  FREE_NOT_NULL(configuration.prefix)
+
+  DEBUG("%s configuration completed.", WRITE_STATSD_NAME);
+  DEBUG("%s Host: %s", WRITE_STATSD_NAME, configuration.host);
+  DEBUG("%s Port: %i", WRITE_STATSD_NAME, configuration.port);
+  DEBUG("%s SilenceTypeWarnings: %i", WRITE_STATSD_NAME,
+        configuration.silence_type_warnings);
+
+  DEBUG("%s FormatString: %s", WRITE_STATSD_NAME,
+        configuration.event_line_format);
+  DEBUG("%s ShortFormatString: %s", WRITE_STATSD_NAME,
+        configuration.short_event_line_format);
+  return 0;
+}
+
+
+static int write_statsd_send_message(const char* message) {
+  int slen;
+  int result;
+  write_statsd_link_t link = open_socket();
+
+  if (link.sock == 0) {
+    return 0;
+  }
+
+  slen = sizeof(link.server);
+  result = sendto(link.sock, message, strlen(message), 0,
+                  (struct sockaddr *)&link.server, slen);
+
+  if (result == -1) {
+    ERROR("write_statsd plugin: unable to send message.");
+  };
+
+  close(link.sock);
+  return 0;
+}
+
+static int write_statsd_write(
+    const data_set_t *ds, const value_list_t *vl, user_data_t *ud) {
+  int   idx;
+  _Bool include_ds_name = configuration.always_append_ds || ds->ds_num > 1;
+
+  /*
+   * Build the type name using all available fields in vl:
+   * hostname.plugin[.plugin_instance].type[.type_instance]
+   */
+  char*  value_key = NULL;
+  size_t value_key_len = 2;  // Dots are always present.
+  size_t plugin_instance_len = strlen(vl->plugin_instance);
+  size_t type_instance_len = strlen(vl->type_instance);
+
+  value_key_len += strlen(vl->host);
```

Comment:
Each part of the name is at most `DATA_MAX_NAME_LEN` bytes long – just use a `char[6 * DATA_MAX_NAME_LEN]` on the stack.